packages/zoe/src/contractFacet/zcfZygote.js

```diff
@@ -326,31 +248,104 @@ export const makeZCFZygote = (
-        reason => {
-          if (reason === undefined) {
-            const newErr = new Error(
-              `If an offerHandler throws, it must provide a reason of type Error, but the reason was undefined. Please fix the contract code to specify a reason for throwing.`,
-            );
-            throw zcfSeat.fail(newErr);
-          }
-          throw zcfSeat.fail(reason);
-        },
-      );
+      const offerResultP =
+        typeof offerHandler === 'function'
+          ? E(offerHandler)(zcfSeat, seatData.offerArgs)
+          : // @ts-expect-error Type issues?
+            E(offerHandler).handle(zcfSeat, seatData.offerArgs);
+
+      const offerResultPromise = offerResultP.catch(reason => {
+        if (reason === undefined) {
+          const newErr = new Error(
+            `If an offerHandler throws, it must provide a reason of type Error, but the reason was undefined. Please fix the contract code to specify a reason for throwing.`,
+          );
+          throw zcfSeat.fail(newErr);
+        }
+        throw zcfSeat.fail(reason);
+      });
       const exitObj = makeExitObj(seatData.proposal, zcfSeat);
       /** @type {HandleOfferResult} */
-      return harden({ offerResultP, exitObj });
+      return harden({ offerResultPromise, exitObj });
     },
   });
+  const handleOfferObj = makeHandleOfferObj();
+
+  const evaluateContract = () => {
+    let bundle;
+    if (passStyleOf(contractBundleCap) === 'remotable') {
+      const bundleCap = contractBundleCap;
+      // @ts-expect-error vatPowers is not typed correctly: https://github.com/Agoric/agoric-sdk/issues/3239
+      bundle = powers.D(bundleCap).getBundle();
+    } else {
+      bundle = contractBundleCap;
+    }
+    return evalContractBundle(bundle);
+  };
+  // evaluate the contract (either the first version, or an upgrade). start is
+  // from a non-upgradeable contract, setupInstallation for upgradeable ones.
+  const { setupInstallation, start, buildRootObject } =
+    await evaluateContract();
+
+  if (start === undefined && setupInstallation === undefined) {
+    assert(
+      buildRootObject === undefined,
+      X`Did you provide a vat bundle instead of a contract bundle?`,
+    );
+    assert.fail(X`unrecognized contract exports`);
+  }
+
+  const contractBaggage = provideDurableMapStore(zcfBaggage, 'contractBaggage');
+
+  // start a non-upgradeable contract
+  const startClassicContract = privateArgs => {
+    /** @type {Promise<ExecuteClassicContractResult>} */
+    const result = E.when(
+      start(zcf, privateArgs),
+      ({
+        creatorFacet = undefined,
+        publicFacet = undefined,
+        creatorInvitation = undefined,
+      }) => {
+        return harden({
+          creatorFacet,
+          publicFacet,
+          creatorInvitation,
+          handleOfferObj,
+        });
+      },
+    );
+    handlePWarning(result);
+    return result;
+  };
+
+  // If the contract is upgradeable, call setupInstallation now.
```

Comment:
Wasn't clear to me at first how you know at this point whether the contract is upgradeable. Contract upgradeable iif setupInstallation is defined?

In that case, maybe this function should talk about "installation setup" rather than upgrade. E.g. `startClassicContract` is `startWithoutInstallation` or `startWithoutSetup`.

Then you don't need this comment line because you don't have to connect `setupInstallation` to the "upgradeable" condition

Reply:
Much simpler now!